velox/exec/tests/utils/QueryAssertions.cpp

```diff
@@ -549,52 +549,310 @@ std::string toString(const MaterializedRow& row) {
+  }
+  for (auto i : floatColumnIndices_) {
+    if (equal(lhs[i], rhs[i])) {
+      continue;
+    }
+    // The 1st non-equal element determines if 'left' is smaller or not.
+    return lessThan(lhs[i], rhs[i]);
+  }
+  return lhs.size() < rhs.size();
+}
 
-  std::vector<MaterializedRow> missingActualRows;
-  std::set_difference(
-      expectedRows.begin(),
-      expectedRows.end(),
-      actualRows.begin(),
-      actualRows.end(),
-      std::inserter(missingActualRows, missingActualRows.end()));
+bool MaterializedRowsForEpsilonComparison::lessThanWithEpsilon(
+    const MaterializedRow& lhs,
+    const MaterializedRow& rhs) {
+  return customLessThan(
+      lhs,
+      rhs,
+      [](const variant& lhs, const variant& rhs) {
+        return lhs.equalsWithEpsilon(rhs);
+      },
+      [](const variant& lhs, const variant& rhs) {
+        return lhs.lessThanWithEpsilon(rhs);
+      });
+}
 
+bool MaterializedRowsForEpsilonComparison::equalWithEpsilon(
+    const MaterializedRow& lhs,
+    const MaterializedRow& rhs) {
+  if (lhs.size() != rhs.size()) {
+    return false;
+  }
+  auto size = lhs.size();
+  for (auto i = 0; i < size; ++i) {
+    if (!lhs[i].equalsWithEpsilon(rhs[i])) {
+      return false;
+    }
+  }
+  return true;
+}
+
+std::string makeErrorMessage(
+    const std::vector<MaterializedRow>& missingRows,
+    const std::vector<MaterializedRow>& extraRows,
+    size_t expectedSize,
+    size_t actualSize) {
   std::ostringstream message;
-  message << "Expected " << expectedRows.size() << ", got " << actualRows.size()
-          << std::endl;
-  message << extraActualRows.size() << " extra rows, "
-          << missingActualRows.size() << " missing rows" << std::endl;
+  message << "Expected " << expectedSize << ", got " << actualSize << std::endl;
+  message << extraRows.size() << " extra rows, " << missingRows.size()
+          << " missing rows" << std::endl;
 
-  auto extraRowsToPrint = std::min((size_t)10, extraActualRows.size());
+  auto extraRowsToPrint = std::min((size_t)10, extraRows.size());
   message << extraRowsToPrint << " of extra rows:" << std::endl;
 
   for (int32_t i = 0; i < extraRowsToPrint; i++) {
     message << "\t";
-    printRow(extraActualRows[i], message);
+    printRow(extraRows[i], message);
     message << std::endl;
   }
   message << std::endl;
 
-  auto missingRowsToPrint = std::min((size_t)10, missingActualRows.size());
+  auto missingRowsToPrint = std::min((size_t)10, missingRows.size());
   message << missingRowsToPrint << " of missing rows:" << std::endl;
   for (int32_t i = 0; i < missingRowsToPrint; i++) {
     message << "\t";
-    printRow(missingActualRows[i], message);
+    printRow(missingRows[i], message);
     message << std::endl;
   }
   message << std::endl;
   return message.str();
 }
 
+std::string MaterializedRowsForEpsilonComparison::getUserFriendlyDiff() {
+  VELOX_CHECK(
+      initialized_,
+      "Object must be initialized via validForEpsilonComparison first.");
```

Comment:
A reference to validForEpsilonComparison needs updating as this method doesn't exist.